### PR TITLE
Accept const session in TF interface.

### PR DESCRIPTION
**Pull request by riga**
#### PR description

This PR adds five new functions to the TensorFlow interface in PhysicsTools that accept `const` sessions, four of which are related to performing model inference and one function handles the session deletion. There's also an additional test case that covers the handling of const sessions.

With this PR, there is no need for them to be stored once per stream module instance, but they can rather be moved to global caches (as is the case for tf graphs at the moment), reducing copies in memory.

The latter changes are not contained in this PR, but are subject to a future one(s). For completeness, in the following there are three lists of files where they'd need to happen (with the latter one being optional) and I can open a dedicated PR for them once _this_ one is merged (or a separate PR per subsystem, depending on what's easier to sign off on).

###### Move sessions to global cache

- L1Trigger/Phase2L1ParticleFlow/src/TauNNId.cc
- RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc
- RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc
- RecoTauTag/RecoTau/plugins/DeepTauId.cc
- RecoTracker/TkSeedGenerator/plugins/DeepCoreSeedGenerator.cc

###### Remove now obsolete `const_cast`'s

- RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc
- RecoHGCal/TICL/plugins/PatternRecognitionbyCLUE3D.cc
- RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.cc
- RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc
- RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc
- RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc
- RecoTracker/MkFit/plugins/MkFitOutputConverter.cc

###### Use const session (only if needed)

- DQM/DTMonitorClient/src/DTOccupancyTestML.cc
- L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc
- L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc
- PhysicsTools/PatAlgos/interface/BaseMVAValueMapProducer.h
- RecoEcal/EgammaCoreTools/src/DeepSCGraphEvaluation.cc
- RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

#### PR validation

I added test cases for handling const sessions.

@valsdav @yongbinfeng @jeongeun @jpata @clacaputo

### CODE REVIEW

File: PhysicsTools/TensorFlow/interface/TensorFlow.h [hunk 1]
```diff
@@ -110,6 +112,15 @@ namespace tensorflow {
            std::vector<Tensor>* outputs,
            const thread::ThreadPoolOptions& threadPoolOptions);
 
+  // version of the function above that accepts a const session
+  inline void run(const Session* session,
+                  const NamedTensorList& inputs,
+                  const std::vector<std::string>& outputNames,
+                  std::vector<Tensor>* outputs,
+                  const thread::ThreadPoolOptions& threadPoolOptions) {
+    run(const_cast<Session*>(session), inputs, outputNames, outputs, threadPoolOptions);
```
[the remark this answers is not in the document]
Comment (riga) [reply PhysicsTools/TensorFlow/interface/TensorFlow.h:123]: Added in b83516cc.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by clacaputo** [comment]
> I can open a dedicated PR for them once _this_ one is merged (or a separate PR per subsystem, depending on what's easier to sign off on)

Hi @riga , thanks a lot. I would suggest using a "per-subsystem" approach for the next PRs

**Comment by riga** [comment]
@clacaputo Ok, will do.

Shall I also adjust the files in the 3 list? In these files, it seems like there is only one session used without concurrency (*), and the change would only consist of making it `const`. However, it could be good to open PRs for these cases anyway, mainly to get confirmation that the assumption (*) is actually correct.

**Comment by clacaputo** [comment]
> Shall I also adjust the files in the third list? In these files, it seems like there is only one session used without concurrency (*), and the change would only consist of making it `const`. However, it could be good to open PRs for these cases anyway, mainly to get confirmation that the assumption (*) is actually correct.

We could directly ping the relevant POG/DPG, or open an issue

**Comment by clacaputo** [comment]
+reconstruction

- TF interface now accepts `const` sessions
- additional PRs will be made. A dedicated issue has been created https://github.com/cms-sw/cmssw/issues/40248
- no reco changes observed

**Pull request merged**